client/gutenberg/editor/controller.js

```diff
@@ -285,3 +286,14 @@ export const exitPost = ( context, next ) => {
 	}
 	next();
 };
+
+export function redirectToPermalinkIfLoggedOut( context, next ) {
+	if ( isUserLoggedIn( context.store.getState() ) ) {
+		return next();
+	}
+	const siteFragment = context.params.site || getSiteFragment( context.path );
+	// Redirect the user to the permalink of the post,page, custom post type if the post is published.
+	// else the endpoint will send the user to the login page.
+	window.location = `https://public-api.wordpress.com/wpcom/v2/redirect?path=${ context.path }&site=${ siteFragment }`;
```

Comment:
I wonder if we could also add another check and do this only for certain post types. We're doing it in the endpoint already, but we could also do it here and save an extra check or redirect if the post type isn't among the supported ones.